File: src/07.greedy/01.arrayGreedy/14.452.ts
export function findMinArrowShots(points: number[][]): number {
    const n = points.length;
    if (n <= 1) {
        return n;
    }

    /**
    // 左端排序
    points.sort((a, b) => a[0] - b[0]);
    let res = 1;
    // 求交集思想
    const range = [...points[0]];

    for (let i = 1; i < n; i++) {
        const item = points[i];
        // 有交集, 就获取其交叉区间
        if (item[0] <= range[1]) {
            range[0] = Math.max(range[0], item[0]);
            range[1] = Math.min(range[1], item[1]);
        } else {
            res++;
            range[0] = item[0];
            range[1] = item[1];
        }
    }

    return res;
     */

    // [[2,5],[1, 6],[7,12],[10,16]]
    // 右端排序
    points.sort((a, b) => a[1] - b[1]);

    // 发射点设置在区间最右侧
    let pos = points[0][1];
    let res = 1;

    for (let i = 1; i < n; i++) {
        if (points[i][0] > pos) {
            pos = points[i][1];
            res++;
        }
    }

    return res;
}

File: src/07.greedy/01.arrayGreedy/14.452.go
package array_greedy

import (
	"sort"
)

func FindMinArrowShots(points [][]int) int {
	n := len(points)
	if n <= 1 {
		return n
	}

	/**
	// 左端排序
	sort.Slice(points, func(i, j int) bool {
		return points[i][0] < points[j][0]
	})

	fmt.Println(points)
	res := 1
	ran := []int{points[0][0], points[0][1]}

	for i := 1; i < n; i++ {
		// 如果有交集
		cur := points[i]

		if cur[0] <= ran[1] {
			if cur[0] > ran[0] {
				ran[0] = cur[0]
			}
			if cur[1] < ran[1] {
				ran[1] = cur[1]
			}
		} else {
			res++
			ran[0] = cur[0]
			ran[1] = cur[1]
		}

	}

	return res
	*/

	// 右端排序
	sort.Slice(points, func(i, j int) bool {
		return points[i][1] < points[j][1]
	})

	res := 1
	pos := points[0][1]

	for i := 1; i < n; i++ {
		cur := points[i]
		if cur[0] > pos {
			pos = cur[1]
			res++
		}
	}

	return res
}
